File: music-catalog-service/src/main/java/io/audira/catalog/dto/ArtistMetricsDetailed.java
package io.audira.catalog.dto;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

/**
 * DTO detallado que representa la evolución temporal de las métricas de un artista.
 * <p>
 * Diseñado para cumplir con el requisito <b>GA01-109: Vista detallada</b>.
 * Contiene tanto los totales para un rango de fechas específico como una lista
 * de puntos de datos diarios ({@link DailyMetric}) optimizada para renderizar
 * gráficos de líneas o barras en el frontend.
 * </p>
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class ArtistMetricsDetailed {

    // --- Contexto del Reporte ---
    /** ID del artista consultado. */
    private Long artistId;

    /** Nombre del artista. */
    private String artistName;

    /** Fecha de inicio del rango del reporte (inclusive). */
    private LocalDate startDate;

    /** Fecha de fin del rango del reporte (inclusive). */
    private LocalDate endDate;

    // --- Datos para Gráficos ---

    /**
     * Lista cronológica de métricas diarias.
     * <p>
     * Cada elemento representa un punto en el eje X (Fecha) con sus valores Y (Plays, Ventas, etc.).
     * Ideal para librerías de gráficos como Chart.js o Recharts.
     * </p>
     */
    private List<DailyMetric> dailyMetrics;

    // --- Resumen del Periodo Seleccionado ---
    /** Suma total de reproducciones dentro del rango [startDate, endDate]. */
    private Long totalPlays;

    /** Suma total de ventas dentro del rango. */
    private Long totalSales;

    /** Ingresos totales generados dentro del rango. */
    private BigDecimal totalRevenue;

    /** Total de comentarios recibidos dentro del rango. */
    private Long totalComments;

    /** Promedio de valoración ponderado durante este periodo. */
    private Double averageRating;

    /**
     * Clase interna estática que representa un punto de datos diario.
     * <p>
     * Se utiliza para granularidad diaria en gráficos de evolución.
     * </p>
     */
    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class DailyMetric {
        /**
         * Fecha del dato (Eje X).
         */
        private LocalDate date;

        /** Cantidad de reproducciones en este día específico. */
        private Long plays;

        /** Cantidad de ventas realizadas en este día. */
        private Long sales;

        /** Ingresos generados en este día. */
        private BigDecimal revenue;

        /** Nuevos comentarios recibidos en este día. */
        private Long comments;

        /**
         * Promedio de valoración calculado solo con los votos de este día.
         * <p>Si no hubo votos, puede ser null o 0.0 dependiendo de la implementación.</p>
         */
        private Double averageRating;
    }
}

File: audira_frontend/lib/core/models/artist_metrics_detailed.dart
import 'package:equatable/equatable.dart';

/// Detailed metrics for an artist with timeline data
/// GA01-109: Vista detallada (por fecha/gráfico básico)
class ArtistMetricsDetailed extends Equatable {
  final int artistId;
  final String artistName;
  final DateTime startDate;
  final DateTime endDate;
  final List<DailyMetric> dailyMetrics;
  final int totalPlays;
  final int totalSales;
  final double totalRevenue;
  final int totalComments;
  final double averageRating;

  const ArtistMetricsDetailed({
    required this.artistId,
    required this.artistName,
    required this.startDate,
    required this.endDate,
    required this.dailyMetrics,
    required this.totalPlays,
    required this.totalSales,
    required this.totalRevenue,
    required this.totalComments,
    required this.averageRating,
  });

  factory ArtistMetricsDetailed.fromJson(Map<String, dynamic> json) {
    return ArtistMetricsDetailed(
      artistId: (json['artistId'] as num).toInt(),
      artistName: json['artistName'] as String,
      startDate: DateTime.parse(json['startDate'] as String),
      endDate: DateTime.parse(json['endDate'] as String),
      dailyMetrics: (json['dailyMetrics'] as List)
          .map((e) => DailyMetric.fromJson(e as Map<String, dynamic>))
          .toList(),
      totalPlays: (json['totalPlays'] as num).toInt(),
      totalSales: (json['totalSales'] as num).toInt(),
      totalRevenue: (json['totalRevenue'] as num).toDouble(),
      totalComments: (json['totalComments'] as num).toInt(),
      averageRating: (json['averageRating'] as num).toDouble(),
    );
  }

  @override
  List<Object?> get props => [
        artistId,
        artistName,
        startDate,
        endDate,
        dailyMetrics,
        totalPlays,
        totalSales,
        totalRevenue,
        totalComments,
        averageRating,
      ];
}

/// Daily metric data point for charts
class DailyMetric extends Equatable {
  final DateTime date;
  final int plays;
  final int sales;
  final double revenue;
  final int comments;
  final double averageRating;

  const DailyMetric({
    required this.date,
    required this.plays,
    required this.sales,
    required this.revenue,
    required this.comments,
    required this.averageRating,
  });

  factory DailyMetric.fromJson(Map<String, dynamic> json) {
    return DailyMetric(
      date: DateTime.parse(json['date'] as String),
      plays: (json['plays'] as num).toInt(),
      sales: (json['sales'] as num).toInt(),
      revenue: (json['revenue'] as num).toDouble(),
      comments: (json['comments'] as num).toInt(),
      averageRating: (json['averageRating'] as num).toDouble(),
    );
  }

  @override
  List<Object?> get props =>
      [date, plays, sales, revenue, comments, averageRating];
}
